ui/src/app/components/error-handler.tsx

```diff
@@ -0,0 +1,132 @@
+import * as React from 'react';
+
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
+
+import {ClrIcon} from 'app/components/icons';
+import {TextModal} from 'app/components/text-modal';
+import {statusApi} from 'app/services/swagger-fetch-clients';
+import colors from 'app/styles/colors';
+import {reactStyles, withGlobalError} from 'app/utils';
+import {globalErrorStore} from 'app/utils/navigation';
+import {ErrorCode, ErrorResponse} from 'generated/fetch';
+import {Button} from './buttons';
+import {Modal, ModalBody, ModalFooter, ModalTitle} from './modals';
+
+const styles = reactStyles({
+  errorHandler: {
+    position: 'fixed',
+    bottom: '1rem',
+    background: colors.white,
+    padding: '0.6rem',
+    border: `1px solid ${colors.black}`
+  },
+  iconStyles: {
+    cursor: 'pointer',
+  },
+  serverStatusList: {
+    margin: '0.5rem 0'
+  }
+});
+
+interface ServerStatus {
+  apiDown: boolean;
+  firecloudDown: boolean;
+  notebooksDown: boolean;
+}
+
+interface Props {
+  globalError: ErrorResponse;
+}
+
+interface State {
+  serverStatus: ServerStatus;
+  serverStatusAcknowledged: boolean;
+}
+
+// We need to maintain this outside of the component, because new errors could come in at any time and we don't want to
+// have it reconstruct/reinitialize with new errors.
+const shouldCheckStatus = new BehaviorSubject<any>(true);
+
+export const ErrorHandler = withGlobalError()(class extends React.Component<Props, State> {
+  obs;
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      serverStatus: {apiDown: false, firecloudDown: false, notebooksDown: false},
+      serverStatusAcknowledged: false
+    };
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    if (this.state.serverStatus !== prevState.serverStatus) {
+      this.setState({serverStatusAcknowledged: false});
+    }
+    const {globalError} = this.props;
+    if (globalError !== prevProps.globalError && globalError !== undefined) {
+      if ((globalError.statusCode === 500
+          || globalError.statusCode === 503)
+          && shouldCheckStatus.getValue()) {
```

Comment:
i am not sure i understand the what shouldCheckStatus is doing, if there is a new globalError with the either code 500 and 503 we set this to false and in pingServer set it to True. Is it like a static variable that is like yes we know there is a service down and we are taking care of it so do not send a new request again to pingServer?

Reply:
Exactly. Otherwise if the api is down, it would get a 500, and decide it should check status, ping again, and infinite loop. This ensures that it only pings once, and doesn't try again while a server check is already in progress.